compiler/src/iree/compiler/Codegen/Common/GenericVectorization.cpp

```diff
@@ -83,13 +86,81 @@ inferVectorSizesFromIR(linalg::LinalgOp linalgOp) {
   return vectorSizes;
 }
 
-// Return the vector sizes from the local lowering config or try to infer them
+static std::optional<SmallVector<int64_t>>
+inferVectorSizesFromIR(tensor::PackOp op) {
+  LLVM_DEBUG(VEC_DBGS() << "Inferring vector sizes for:\n" << op << "\n");
+
+  if (llvm::any_of(op.getInnerTiles(), [](OpFoldResult v) {
+        return !getConstantIntValue(v).has_value();
+      })) {
+    LLVM_DEBUG(VEC_DBGS() << "skip, because inner_tiles are not all constant");
+    return std::nullopt;
+  }
+
+  SmallVector<int64_t> vectorSizes;
+  auto inferred = inferVectorSizesFromIR(op.getSource());
+  if (!inferred) {
+    return std::nullopt;
+  }
+  vectorSizes = inferred.value();
+
+  for (auto [dimPos, tileSize] :
+       llvm::zip_equal(op.getInnerDimsPos(), op.getStaticInnerTiles())) {
+    if (vectorSizes[dimPos] % tileSize != 0) {
+      return std::nullopt;
+    }
+    vectorSizes[dimPos] /= tileSize;
+  }
+  auto outerDimsPerm = op.getOuterDimsPerm();
+  if (!outerDimsPerm.empty()) {
+    applyPermutationToVector(vectorSizes, outerDimsPerm);
+  }
+
+  LLVM_DEBUG({
+    VEC_DBGS() << "After adjustment with inner tiles and "
+                  "outer_dims_perm:\n";
+    for (auto [idx, val] : llvm::enumerate(vectorSizes)) {
+      llvm::dbgs() << "Dim #" << idx << ": " << val << "\n";
+    }
+  });
+
+  return vectorSizes;
+}
+
```

Comment:
Sorry, I'm a bit confused here. Perhaps some doc for the different `inferVectorSizesFromIR` versions would help. Why do we need all these per op specializations? It looks like we have to adjust the inferred vector sizes, which make me wonder if the inference using value bounds also fall short for these cases.

Reply:
> Why do we need all these per op specializations?

All these operations implement their own tiling interface methods and vectorization, so it makes sense to have their own vector input size inference.

> It looks like we have to adjust the inferred vector sizes, which make me wonder if the inference using value bounds also falls short for these cases.

The adjustment is because of pack/unpack op semantic. The vector_input_size/tiling sizes are all about destination shape. Let's take a look at pack op. If the shapes are inferred from source (i.e., non-packed domain), we need to adjust them to get the sizes in packed domain. In the past implementation, it only looks at operands that are tensor.extract_slice ops. The inference could fail for `linalg -> pack`. With the PR, we look up through the chain, so it is more robust for inference.

Reply:
After thinking more, I found that there are actually two sizes that I mixed them in the previous revision. One is result shape and the other is input vector sizes. We need to follow the operation chain to infer result shape, and use them to get input vector sizes. They are the same in most of cases (e.g., tensor.pack), but they are needed for a LinalgOp. Because it could be reductions and the consumers only care about the corresponding OpResult shapes. I added a struct and more documentation. Please take a look, thanks!